### ISSUE TITLE
Colorama subdependency preventing from deploying microservice on docker

### ISSUE THREAD

**Issue by brupelo** [pypa/pipenv#5085]
Hi guys,

I'd like some clarification over here as I'm slightly confused, few hours ago I've opened this [issue](https://github.com/pallets/click/issues/2261) on click issue tracker because click is marking one of its subdependencies as windows specific (even if that colorama subdependency is cross-platform) producing that a flask/pyenv microservice can't be deployed out of the box unless we add explicitely that colorama subdependency as `colorama = {version = "*", sys_platform="!='fake'"}` in the Pipfile

I'm wondering, is this a pipenv bug or intended behaviour? Over here we're still trying to decide which solution (pipenv, poetry, pdm) to adopt for our pipelines/deployments and from what I've heard poetry it does produces cross-platform lock files while that's not being the case with pipenv.

Could you please clarify if this is intended behaviour or not? And if this is indeed a bug, is there any plan to address that?

Thanks in advance.

**Comment by brupelo** [comment]
@matteius First of all, thanks to jump on this one and sorry for not having provided proper details at the very beginning so let me try again.

I've cloned pipenv repo and checkout main branch at ec34097407d0f8a98a12e6a9ae2caf32d76de4cd :

```
d:\mcve_pipenv>pipenv --version
pipenv, version 2022.4.22.dev0

```

Then I've just reduced my original microservice Pipfile to the bare minimum:

```
d:\mcve_pipenv>cat Pipfile
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
# colorama = {version = "*", sys_platform="!='fake'"}
Flask-Cors = "3.0.10"

```
So if I run `pipenv install` I'll get a Pipfile lock that contains:

```
d:\mcve_pipenv>python -c "import json; from pathlib import Path;print(json.loads(Path('Pipfile.lock').read_text())['default']['colorama'])"
{'hashes': ['sha256:5941b2b48a20143d2267e95b1c2a7603ce057ee39fd88e7329b0c292aa16869b', 'sha256:9f47eda37229f68eee03b24b9748937c7dc3868f906e8ba69fbcbdd3bc5dc3e2'], 'markers': "platform_system == 'Windows'", 'version': '==0.4.4'}

```

So, could you guys please clarify if this is intended behaviour or a bug? Basically I guess what I want to know whether if it's possible to generate cross-platform `Pipfile.lock` using pipenv or not, as we've got basically developers using windows/linux/macosx, even if in our pipelines/deployments we'll be using mostly docker/ubuntu.

Thanks for looking into this one.

**Comment by matteius** [comment]
@brupelo Thanks for your example -- I just tried this on Windows vs Linux with the latest version and you are correct that on linux  the colorama dependency is not included in the Lockfile.  Perhaps it is a potential bug to do with transitive dependencies? 

When including it in the Pipfile as well as `colorama = {version = "*"}` it is included for all systems without the windows marker.  

If I include it as `colorama = {version = "*", sys_platform="=='windows'"}`  Then the lockfile on linux is generated to include:
```
        "colorama": {
            "hashes": [
                "sha256:5941b2b48a20143d2267e95b1c2a7603ce057ee39fd88e7329b0c292aa16869b",
                "sha256:9f47eda37229f68eee03b24b9748937c7dc3868f906e8ba69fbcbdd3bc5dc3e2"
            ],
            "markers": "sys_platform == 'windows'",
            "version": "==0.4.4"
        },
```

So it is possible to generate a cross platform lockfile as long as the cross platform dependencies are called out in the Pipfile.  It may be a bug that the transitive dependencies are only included for the matching system platform that the lock file is generated on, unfortunately.    We are a patch friendly community though if you or someone else wants to work on improving it, I can try to point you at some places to start your investigation.  Anyway it seems the current workaround in pipenv is to include those system specific transitive dependencies in the top level Pipfile with version '*' and the version specifiers in the packages that require them will take effect and you'll get the desired outcome in the current versions, with that extra bit of definition.

**Comment by matteius** [comment]
@brupelo Was my last response received, did it make sense?  I'd like to determine what to do with this issue report, is there an actionable item or was it more of a question about the current behavior?  It seems there are some other tickets opened recently around enhancements for cross platform lockfiles.  Maybe tracking this is sufficient?   https://github.com/pypa/pipenv/issues/5130
